graylog2-web-interface/src/views/components/MigrateFieldCharts.jsx

```diff
@@ -0,0 +1,189 @@
+      break;
+    case 'mean':
+      seriesName = 'avg';
+      break;
+    case 'cardinality':
+      seriesName = 'card';
+      break;
+    default:
+      seriesName = legacySeriesName;
+  }
+  return `${seriesName}(${field})`;
+};
+
+const mapVisualization = (visualization: LegacyVisualization) => {
+  switch (visualization) {
+    case 'scatterplot':
+      return 'scatter';
+    default:
+      return visualization;
+  }
+};
+
+const createVisualizationConfig = (interpolation: LegacyInterpolation, visualization: string) => {
+  let interpolationName: InterpolationMode;
+  switch (interpolation) {
+    case 'basis':
+    case 'bundle':
+    case 'cardinal':
+    case 'monotone':
+      interpolationName = 'spline';
+      break;
+    case 'linear':
+    case 'step-after':
+      interpolationName = interpolation;
+      break;
+    default:
+      interpolationName = 'linear';
+  }
+
+  switch (visualization) {
+    case 'line':
+      return new LineVisualizationConfig(interpolationName);
+    case 'area':
+      return new AreaVisualizationConfig(interpolationName);
+    default:
+      return undefined;
+  }
+};
+
+const onMigrate = (legacyCharts: Array<LegacyFieldChart>, setMigrating: boolean => void) => {
+  const widgetDef = widgetDefinition('AGGREGATION');
+  const currentView = CurrentViewStateStore.getInitialState();
+  const newWidgetPositions = { ...currentView.state.widgetPositions };
+  const lastWidgetPosition = maxBy(Object.values(newWidgetPositions), (position: WidgetPosition): number => position.row);
+
+  const newWidgets = legacyCharts.map((chart: LegacyFieldChart, index: number) => {
+    setMigrating(true);
+
+    const { field } = chart;
+    // The old field charts only have one series per chart.
+    const series = new Series(mapSeries(chart.valuetype, field));
+    // Setting series. The old field charts only have one series per chart.
+    const rowPivots = [new Pivot('timestamp', 'time', { interval: { type: 'timeunit', ...mapTime(chart.interval) } })];
+    const visualization = mapVisualization(chart.renderer);
+    const visualizationConfig = createVisualizationConfig(chart.interpolation, visualization);
+    const widgetConfig = AggregationWidgetConfig.builder()
+      .visualization(visualization)
+      .visualizationConfig(visualizationConfig)
+      .series([series])
+      .rowPivots(rowPivots)
+      .build();
+    const newWidget = AggregationWidget.builder()
+      .newId()
+      .timerange(undefined)
+      .config(widgetConfig)
+      .build();
+    const widgetRow = lastWidgetPosition.row + lastWidgetPosition.height + index + widgetDef.defaultHeight;
+    newWidgetPositions[newWidget.id] = new WidgetPosition(1, widgetRow, widgetDef.defaultHeight, Infinity);
+    return newWidget;
+  });
+
+  const newViewState = currentView.state
+    .toBuilder()
+    .widgets(Immutable.List([...currentView.state.widgets, ...newWidgets]))
+    .widgetPositions(newWidgetPositions)
+    .build();
```

Comment:
When writing the tests I mocked `CurrentViewStateStore.getInitialState` to return an object with widgets and widget positions. This way I can't use currentView.state.toBuilder() because it is not a function. It is also not possible to use `ViewState.create` inside the mock. Does this mean I have to create the current view + its state manually before running the tests? Or can you think of a better way?

Reply:
Answer: With `useMock` we can return the `ViewState` inside the `CurrentViewStateStore.getInitialState` mock and don't have to create the full View initally.